lib/ddtrace/configuration/components.rb

```diff
@@ -106,6 +109,33 @@ def build_tracer_options(settings)
           end
         end
 
+        def build_tracer_test_mode_options(settings)
+          {}.tap do |opts|
+            # Enable partial flushing:
+            #   When in test mode, traces won't complete until tests complete.
+            #   We need to flush spans from the context before its finished so
+            #   tests can assert upon trace output, if needed.
+            opts[:partial_flush] = true
+            # NOTE: Setting threshold to 1 makes this like "span streaming"
+            #       A side effect of doing so is every span will receive both
+            #       "origin" and "sampling_priority" tags. This is expected,
+            #       as currently "origin" needed on every test span.
+            opts[:min_spans_before_partial_flush] = 1
+
+            # Temporary adjustment for flushing test spans:
+            # Flushes spans more aggressive from the buffer.
+            # Replace this with span streaming or synchronous writing.
```

Comment:
In these test environments, is the trace-agent running in [syncMode](https://github.com/DataDog/datadog-agent/blob/140a4ee164261ef2245340c50371ba989fbeb038/pkg/trace/writer/trace.go#L65) as well, ala how it runs in Lambda? I do think that sync writing makes sense to work toward if we need high guarantees that all spans flush.

Reply:
Right now, no. But it could, potentially.

Downside with synchronous writing is that it doesn't play nicely with some anticipated use.

I would like users to be able to assert things about traces produced within their tests (e.g. "I expected this web request to produce a trace with a Rack span") This would be especially useful for users who have defined manual instrumentation and want to verify its correctness.

If users use the RSpec or Cucumber integration to trace their CI tests, the root span of their traces will be the span for the test (e.g. RSpec span.) Because they would be running assertions on trace output while the test is still running, the trace would not be complete and not written to the writer's trace buffer. Any assertions on trace output will fail, because the trace will not be in the buffer to assert against.

This is where partial flush of size 1 comes in: it would "stream" each span from the context immediately to the writer's buffer, meaning you could assert on written spans. It also has the bonus of tagging every span with "origin", something we currently require of all spans generated during a CI test.

Sync writing doesn't play well with this, because if we activated it with partial flushing size 1, it would do a blocking HTTP request for every span in the test... this would be very bad.

There may be a better strategy to get these two to play together, but I don't have one at the moment. If you have suggestions I'd be happy to hear them.

Reply:
Yes i think sync mode would mean moving away from how we currently do partial flushing, we would instead be queuing many traces/spans in memory (perhaps >10k), and then doing a few large flushes. I would have to review how the agent handles syncronous flushing, ideally we could model our approach after theirs.

Reply:
I think this might be worth engaging with more, now, if there's some reasonable, effective change we can make. I'd rather not mess with buffer/interval settings and expose a public API for it if it isn't comprehensive and there's a better alternative readily available.

I will consider alternate strategies, and share my findings.

Reply:
I decided to go a different direction, based on the feedback.

By default, CI will turn on a SyncWriter & a custom context flush (same as the default, but adds tags.) Partial flushing no longer applies. This should help improve the guarantee that all spans are flushed without having conflicts with partial flush.

Let me know what you think.

Reply:
I think this makes sense. Am I right in saying that the `context_flush` bit about adding tags is not yet implemented?

I think a combination of `context_flush` + `SyncWriter` are correct here. The `test_mode` i'd be curious to see whether users find it to be useful, i know there have been a lot of requests for it, so that's a nice bonus we can get out of this work.

Reply:
I think this comment thread can be resolved when you feel appropriate